### ISSUE TITLE
Issue discovered with ruby-cbc

### ISSUE THREAD

**Issue by tepperly** [gverger/cbc-wrapper#2]
I discovered an issue while trying to use ruby-cbc. Here is the [initial report.](https://github.com/gverger/ruby-cbc/issues/14)  It appears to be an issue with cbc-wrapper.
```sh
$ irb
irb(main):001:0> require "cbc-wrapper"
=> true
irb(main):002:0> Cbc_wrapper.Cbc_newModel
1 was provided for  - valid range is 0 to 0
1e+08 was provided for  - valid range is 0 to 0
5 was provided for  - valid range is 0 to 0
2147483647 was provided for  - valid range is 0 to 0
5 was provided for  - valid range is 0 to 0
1e-06 was provided for  - valid range is 0 to 0
1e-05 was provided for  - valid range is 0 to 0
-1 was provided for  - valid range is 0 to 0
1003 was provided for  - valid range is 0 to 0
1057 was provided for  - valid range is 0 to 0
1 was provided for  - valid range is 0 to 0
-1 was provided for  - valid range is 0 to 0
100 was provided for  - valid range is 0 to 0
=> #<SWIG::TYPE_p_void:0x005582e0fd5288 @__swigtype__="_p_void">
```
Here are some of the details about the machine I am report the issue for:
Linux faerun 4.3.0-1-amd64 #1 SMP Debian 4.3.5-1 (2016-02-06) x86_64 GNU/Linux
$ ruby --version
ruby 2.3.3p222 (2016-11-21) [x86_64-linux-gnu]
I installed ruby-cbc and cbc-wrapper using "gem install ruby-cbc".
$ gem -list | grep cbc
cbc-wrapper (2.9.9.2)
ruby-cbc (0.3.16)
I installed cbc-wrapper via "gem install ruby-cbc".

**Comment by gverger** [comment]
Thanks for moving this issue here.
I'll try and have a look at that probably in the weekend. I am afraid I don't understand what is happening here...

**Comment by tepperly** [comment]
FYI, I am trying to build cbc-wrapper from a git-clone of the repo.  I have coinor-cbc installed via apt-get install.  I double checked with a gdb --pid that it is using the cbc libraries that the cbc-wrapper build built.

**Comment by tepperly** [comment]
Is there an easy way to force it to build against the system copy of cbc?

**Comment by gverger** [comment]
There is not. You can look at what is happening here: https://github.com/gverger/cbc-wrapper/blob/master/ext/cbc-wrapper/extconf.rb

**Comment by tepperly** [comment]
I've looked into this from a bunch of different angles. I wrote a C++ program that did nothing but call Cbc_newModel(), and it worked just fine. I ran it in valgrind, and nothing went wrong.
I do see some stuff that concerns me in the Cbc source code. For example from Cbc_C_Interface.h,
```C++
    Cbc_Model * model = new Cbc_Model();
    OsiClpSolverInterface solver1;
    model->solver_    = &solver1;
```
solver1 is on the stack, and the model pointer is what's returned to the client. Of course, it dies before this is the problem.

**Comment by tepperly** [comment]
What version of ruby are you using?

**Comment by tepperly** [comment]
For what it's worth, I tried compiling ruby-2.4.2 from source and then running "gem install ruby-cbc". It produces the same outcome.

**Comment by gverger** [comment]
Thanks for diving this hard into it!

We use ruby-cbc in my company. It was working from ruby 2.0 to ruby 2.3.5 now.
FYI it is working properly on my mac (osx  from `el capitan` to `sierra`), and it is currently running in Heroku on an Ubuntu 14.04.

**Comment by tepperly** [comment]
I use Cbc for ham radio contest scoring. I have a Ruby contest scoring program, and I currently do the Cbc solve manually. It would be useful to be able to solve it without leaving Ruby.

**Comment by gverger** [comment]
Mmm. As a workaround, could you consider using the `model.to_s` from `ruby-cbc`, which outputs the model in a format readable by Cbc, and then use a `system("cbc model_file.lp")` to finally parse the result in the output?

**Comment by tepperly** [comment]
I've already got the code to make an input file and to parse the output file. I was hoping for something cleaner.

**Comment by gverger** [comment]
It seems Clp is compiled without the `COIN_HAS_CBC` preprocessor feature on. See in `Clp/src/CbcOrClpParam.cpp` in coinor. It then doesn't define all parameters. That is why we see `valid range is 0 to 0`.

I tried adding `CPPFLAGS=-DCOIN_HAS_CBC` in the configure command (`extconf.rb` file), but then it doesn't compile.

There should be a way of making sure the compilation knows that we are building Cbc.

**Comment by gverger** [comment]
It is strange, `configure` sees that Cbc is in the sources (variables `$coin_has_cbc` and `$coin_reason` are `yes` and `source in Cbc`.

![1__bash___users_guillaume__docker_](https://user-images.githubusercontent.com/6671307/32639495-17c97338-c579-11e7-9af9-19d394fc23f9.png)

**Comment by tepperly** [comment]
I tried building against the Debian unstable build, and the same thing
happens.

**Comment by tepperly** [comment]
What is the default compiler for machines where ruby-cbc and cbc-wrapper
are working?

**Comment by gverger** [comment]
On mac osx, we are using clang, and on other machines, it is using g++/gcc

**Comment by tepperly** [comment]
do you know what particular version of g++/gcc?  I am trying to see if its
a compiler thing.

Tom

**Comment by gverger** [comment]
Version of gcc and g++ is 4.8.4

```
$ g++ --version
==> g++ (Ubuntu 4.8.4-2ubuntu1~14.04.3) 4.8.4
$ gcc --version
==> gcc (Ubuntu 4.8.4-2ubuntu1~14.04.3) 4.8.4
```

**Comment by gverger** [comment]
It is working on Ubuntu 17.04 with g++ 6.3.0 too

**Comment by soutaro** [comment]
Is there any update on this issue?

I have the same problem on CentOS7, ruby-2.5.1, gcc 4.8.5.

```
# uname -a
Linux b7ee37e4c53f 4.9.93-linuxkit-aufs #1 SMP Wed Jun 6 16:55:56 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

# ruby -v   
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]

# gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-28)
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by gverger** [comment]
I am sorry, I cannot figure out what is happening.

I created a branch [dyn-lib](https://github.com/gverger/cbc-wrapper/tree/dyn-lib), where we don't compile the cbc library, but use the one which is already present.

For Centos, there is no package for cbc (see [here](https://projects.coin-or.org/Cbc)). I compiled the sources and I have the `cbc` commandline running on a centos. Unfortunately, when using with this gem, still the same problem.

I have no idea what I am missing here...

**Comment by soutaro** [comment]
Thanks!

I tried the `dyn-lib` branch on CentOS using the [distributed binary](https://bintray.com/coin-or/download/Cbc/), and got the same problem too.

The easiest solution for this issue should be switching from CentOS to Ubuntu...

**Comment by gverger** [comment]
Would you mind trying the [ffi](https://github.com/gverger/ruby-cbc/tree/ffi) branch on `ruby-cbc` ?

In this branch, I try to use FFI instead of SWIG to handle the connection to the native library, and it seems to work on my instance of Centos.

This is highly experimental:
- In this version, I got rid of the dependency to `cbc-wrapper`, but in the long term I want to separate wrapping the gem and compiling the lib from the logic in `ruby-cbc`.
- It will only work if you installed `cbc` on your system

Next steps for me are to move the code relying on ffi back in this project, but it will take a bit of time since there are some highly coupled code there I need to decouple.

**Comment by soutaro** [comment]
I got it!

When we link with both `CbcSolver` and `ClpSolver`, it prints the message! The required library is only `CbcSolver`.
I have some fix on the `extconf.rb`, so will open a pull request for `dyn-lib` branch soon.

**Comment by soutaro** [comment]
Probably we don't need to use the system installed cbc.

**Comment by gverger** [comment]
Thank you very much! This problem is now resolved, version 2.9.9.3 fixes it.
